src/small_version.rs

```diff
@@ -2,87 +2,247 @@ use std::sync::Arc;
+///
+/// The exact set of versions that are common enough to get a small representation depends on the size of a pointer
+/// and is subject to change between releases.
+#[derive(Debug, Eq)]
+#[repr(packed)]
+pub struct SmallVersion(*const semver::Version);
 
-#[derive(Debug, PartialEq, Eq, Clone, Hash)]
-enum Iner {
-    Full(Arc<semver::Version>),
-    Small(Small),
+/// polyfill for `std::ptr::without_provenance`
+pub fn without_provenance(addr: usize) -> *const semver::Version {
+    std::ptr::null::<semver::Version>().with_addr(addr)
 }
 
-#[derive(Debug, PartialEq, Eq, Clone, Hash)]
-struct Small {
-    major: u16,
-    minor: u16,
-    patch: u16,
-    pre: bool,
+#[derive(Debug, PartialEq, Eq, Clone, Hash, PartialOrd, Ord)]
+struct Small(usize);
+
+impl SmallVersion {
+    fn new_full(v: semver::Version) -> Self {
+        // Safety: Remember that owning a `SmallVersion` that is a `Full` is equivalent to owning an `Arc`.
+        let out = Self(Arc::into_raw(Arc::new(v)));
+        // Safety: We always check that a newly constructed `SmallVersion` has been tagged correctly
+        // before we return it to code outside this module.
+        assert!(out.is_full());
+        out
+    }
 }
 
-impl Small {
-    fn new(v: &semver::Version) -> Option<Small> {
-        if !v.build.is_empty() {
-            return None;
-        }
-        Some(Small {
-            major: v.major.try_into().ok()?,
-            minor: v.minor.try_into().ok()?,
-            patch: v.patch.try_into().ok()?,
-            pre: if v.pre.is_empty() {
-                false
-            } else if v.pre.as_str() == "0" {
-                true
-            } else {
-                return None;
-            },
-        })
+impl Drop for SmallVersion {
+    fn drop(&mut self) {
+        if let RefIner::Full(ptr) = RefIner::from(&*self) {
+            // Safety: We are a `Full` and so where constructed with `Arc::into_raw`,
+            // and we are being droped, so we must `Arc::from_raw` to drop it.
+            // All notes on `Arc::from_raw` are trivially satisfied because we are not doing any type punning.
+            unsafe { Arc::from_raw(ptr) };
+        }
     }
+}
 
-    fn pre(&self) -> &str {
-        if self.pre {
-            "0"
+impl Clone for SmallVersion {
+    fn clone(&self) -> Self {
+        if let RefIner::Full(ptr) = RefIner::from(self) {
+            // Safety: We are a `Full` and so where constructed with `Arc::into_raw`,
+            // and we are being cloned, so we must increment the strong count to match.
+            // We know that the ptr is still valid, because we have a reference to self.
+            unsafe { Arc::increment_strong_count(ptr) };
+        }
+        Self(self.0)
+    }
+}
+
+impl From<semver::Version> for SmallVersion {
+    fn from(v: semver::Version) -> Self {
+        match (&v).try_into() {
+            Ok(Small(s)) => {
+                let out = SmallVersion(without_provenance(s));
```

Comment:
This is missing a `// SAFETY` comment — why does `without_provenance(s)` conform to the invariants of `SmallVersion`? (This is the sort of missed safety comment that the aforementioned `def` trick helps avoid.)

Reply:
Odd that [std::without_provenance](https://doc.rust-lang.org/std/ptr/fn.without_provenance.html) is safe, given that it should have safety comments.

Reply:
It's the `SmallVersion(...)` invocation that's conceptually unsafe — its field has a safety invariant that must be upheld. The `without_provenance(s)` invocation is fine.

Reply:
perhaps split it over two lines:
```suggestion
                let raw = without_provenance(s);
                // SAFETY: the least significant bit of `raw` is `1`. (but say more about why this is true)
                let out = SmallVersion(raw);
```